Subject: Fix pkg-config on windows

1. Store the PKG_CONFIG in Makefile.sub (or try to get it from the ENV var PKG_CONFIG
2. Pass --msvc-syntax. This assumes you use pkgconf (1.4.0 or above, 7 years ago!) and not pkg-config (freedesktop).
3. try_ldflags is broken, it passes them as "opt" and NOT as ldflags. As a result they (specifically the /libpath:xxx one) end up passed before the `-link` command to CL.exe and it throws because it ignores it and therefore can't find libffi.
```
cl : Command line warning D9002 : ignoring unknown option '-libpath:C:/Users/julien/.conan2/p/libff3726d89a6255c/p/lib'
```

diff --git a/lib/mkmf.rb b/lib/mkmf.rb
--- lib/mkmf.rb
+++ lib/mkmf.rb
@@ -558,7 +558,15 @@ MSG
 
   def try_link0(src, opt="", *opts, &b) # :nodoc:
     exe = CONFTEST+$EXEEXT
-    cmd = link_command("", opt)
+    # Retrieve ldflags from opts
+    ldflags = ""
+    opts.each do |h|
+      if h.is_a?(Hash) && h.key?(:ldflags)
+        ldflags = h[:ldflags]
+        break
+      end
+    end
+    cmd = link_command(ldflags, opt)
     if $universal
       require 'tmpdir'
       Dir.mktmpdir("mkmf_", oldtmpdir = ENV["TMPDIR"]) do |tmpdir|
@@ -678,9 +686,10 @@ MSG
     $LDFLAGS = ldflags unless ret
   end
 
-  def try_ldflags(flags, opts = {})
+  def try_ldflags(ldflags, opts = {})
+    opts = {:ldflags => ldflags}.update(opts)
     opts = {:werror => true}.update(opts) if $mswin
-    try_link(MAIN_DOES_NOTHING, flags, opts)
+    try_link(MAIN_DOES_NOTHING, "", opts)
   end
 
   def append_ldflags(flags, *opts)
@@ -1883,11 +1892,34 @@ SRC
         pkgconfig = nil
       end
       if pkgconfig
+        has_ms_win_syntax = false
+        if $mswin
+          has_ms_win_syntax = xpopen([pkgconfig, "--help"]).read.include?('msvc-syntax')
+          if has_ms_win_syntax
+            args << "--msvc-syntax"
+          else
+            Logging.message("WARNING: #{pkgconfig} does not support the --msvc-syntax. Try using a recent pkgconf instead")
+          end
+        end
         get = proc {|opts|
           opts = Array(opts).map { |o| "--#{o}" }
           opts = xpopen([*envs, pkgconfig, *opts, *args], err:[:child, :out], &:read)
           Logging.open {puts opts.each_line.map{|s|"=> #{s.inspect}"}}
-          opts.strip if $?.success?
+          if $?.success?
+            opts = opts.strip
+            if $mswin and not has_ms_win_syntax
+              opts = Shellwords.shellwords(opts).map { |s|
+                if s.start_with?('-l')
+                  "#{s[2..]}.lib"
+                elsif s.start_with?('-L')
+                  "/libpath:#{s[2..]}"
+                else
+                  s
+                end
+              }.quote.join(" ")
+            end
+            opts
+          end
         }
       end
       orig_ldflags = $LDFLAGS
diff --git a/win32/Makefile.sub b/win32/Makefile.sub
--- win32/Makefile.sub
+++ win32/Makefile.sub
@@ -1131,6 +1131,7 @@ s,@top_srcdir@,$(srcdir),;t t
 s,@try_header@,try_compile,;t t
 s,@ruby_pc@,$(ruby_pc),;t t
 s,@RJIT_SUPPORT@,$(RJIT_SUPPORT),;t t
+s,@PKG_CONFIG@,$(PKG_CONFIG),;t t
 <<KEEP
 
 !if "$(HAVE_BASERUBY)" != "yes" || "$(CROSS_COMPILING)" == "yes"
